design/StableMemory.md

```diff
@@ -0,0 +1,88 @@
+# Stable Memory API
+
+The current implementation of stable variables is based on
+serialization and deserialization of all stable data on upgrade. This
+clearly doesn't scale to large amounts of stable data as there may not
+be enough cycles to perform (de)serialization.
+
+To avoid this upgrade hazard, some Rust canisters with low-level API
+access, and large stable memory footprints arrange to store their
+persistent data in stable memory at all times, using either a custom
+binary encoding or a mixture of candid and raw binary.
+
+To provide more fine-grained access to stable memory we propose:
+
+A) extending Motoko with access to (monomorphic) Candid serialization primitives.
+   This is most easily achieved by extending the syntax with saturated serialization and deserialization
+   constructs, along the lines of debug_show.
+
+```
+   encode<T>(v : T) : Blob
+   decode<T>(b : Blob) : T
```

Comment:
Should this return ?T. I'd want to be able to handle deserialization failure.

Reply:
I'm not sure we have that option with the current serialization code. Even if we did, we have to trust that the contents actually has type `T` I think, but maybe I'm mistaken. @rossberg?

Reply:
Hm, wouldn't these signatures depend on adding type-passing polymorphism first? Short of that, I suspect we can only support monomorphic functions.

And yeah, we probably need to make these return options. I assume that's can be made to work with the deserialisation code, since it already has a mechanism for abort and fall back for mismatches on option types.

Reply:
Sure they should probably just be overloaded constructs applicable to any T  where T shared (just like  equality and similar to debug_show). Just sketching here.